sklearn/cluster/dbscan_.py

```diff
@@ -118,18 +125,30 @@ def dbscan(X, eps=0.5, min_samples=5, metric='minkowski',
     if metric == 'precomputed':
         D = pairwise_distances(X, metric=metric)
         neighborhoods = np.empty(X.shape[0], dtype=object)
-        neighborhoods[:] = [np.where(x <= eps)[0] for x in D]
+        if sparse.issparse(D):
+            D_mask = D.data <= eps
+            masked_indices = D.indices.astype(np.intp)[D_mask]
```

Comment:
For very large matrices, scipy can use 64 bit int for the indices that is `np.intp` on a 64 bit platform. It would therefore be recommend to use `sklearn.utils.fixes.astype(D.indices, copy=False)` here.